Storage/src/Bucket.php

```diff
@@ -231,10 +231,15 @@ public function exists()
      *           of type string or null.
      *     @type bool $resumable Indicates whether or not the upload will be
      *           performed in a resumable fashion.
-     *     @type bool $validate Indicates whether or not validation will be
-     *           applied using md5 hashing functionality. If true and the
-     *           calculated hash does not match that of the upstream server the
-     *           upload will be rejected.
+     *     @type bool|string $validate Indicates whether or not validation will
+     *           be applied using md5 or crc32 hashing functionality. If
+     *           enabled, and the calculated hash does not match that of the
+     *           upstream server, the upload will be rejected. Provide `true` or
+     *           `md5` for md5 hashing, `crc32` for crc32 hashing, or `false` to
+     *           disable. Please note that crc32 may have negative performance
+     *           implications in older versions of PHP. Installation of the
+     *           `crc32c` PHP extension is strongly advised for best
+     *           performance. **Defaults to** `true` (md5).
```

Comment:
@jdpedrie given performance is improved with crc32c, why not set it as the default instead of md5?

Reply:
@frankyn crc32 has serious performance drawbacks in situations where the crc32c extension is not installed and where the php installation does not support it out of the box. In other words, when we have to fall back to the pure-PHP implementation, we see unacceptable performance losses:

The first and second rows below are pure-PHP implementations. The third and fourth are using the extension and built-in implementations.

```
Google\CRC32\PHP	        256	24267       1.24 MB/s
Google\CRC32\PHPSlicedBy4	256	24707       1.26 MB/s
Google\CRC32\Builtin	    256	1582535     81.03 MB/s
Google\CRC32\Google	        256	1812113     92.78 MB/s
```

This corresponds to roughly 87 seconds for a 130mb file in our testing.

@dwsupplee and I were talking about being a bit smarter when choosing a validation algorithm. In situations where the user does not explicitly choose one or the other, I think it should be possible to determine the best option. In PHP 7.4 and up, or when the crc32c extension is installed, we can default to crc32 without losing performance. In other cases, we can default to md5.

WDYT?

(original was replying to the wrong thing)

Reply:
Thank you for running performance analysis, could you add what each column represents? I'm not following the second and third columns. 

How deterministic is the check for PHP version and extension availability? Thank you for describing this issue in the documentation I missed it.

I want to say, "yes we should add a smart selection for md5, crc32c validation", but that will change the behavior expectations of the library on uploads. An alternative solution could be to add a new option called "auto", but we can shelf that until a later time if we receive feedback on it.

Reply:
Sure thing, sorry. 2nd column is the chunk size. 3rd column is the total amount of data hashed in bytes.

It's an easy check. Since the [php-crc32](https://github.com/google/php-crc32) library provides methods to determine whether an implementation is supported.

```php
use Google\CRC32\Builtin;
use Google\CRC32\CRC32;

$supported = function_exists('crc32c') || Builtin::supports(CRC32::CASTAGNOLI);
```

Reply:
I had begun drafting the changes to autoselect a signing method. It works on the assumption that the default is to autoselect, but could easily be changed later to work with a potential `auto` option. If you're curious about how the logic would work, you can take a look at [this commit](https://github.com/jdpedrie/google-cloud-php/commit/11e547362d4d5e4cc41941c219deb5c5d52b7f4a) (not in this pull request)

Reply:
Thanks @jdpedrie, do hashed bytes represent how many bytes are hashed in 87 second?

I like the idea of using `true` to determine what's best for your environment. It reduces overhead for the user. Given the example PR you provided, I'd expect to not see `auto` and instead rely on `true` for auto selection.

I'm guessing you'll move to complete this PR and then follow-up with the autoselect update in a separate PR. Is this correct?

Reply:
> Thanks @jdpedrie, do hashed bytes represent how many bytes are hashed in 87 second?

@frankyn I'm sorry, I misread the benchmark script before. The third column is the number of times `hash_update()` is called in the test.

The 87 seconds number comes from the total time for benchmarking of hashing a 130mb file. We ran some manual tests to see what it looked like for a large file.

The table comes from the php-crc32 automated benchmark, which you can review [here](https://github.com/google/php-crc32/blob/master/crc32_benchmark.php).

> I'm guessing you'll move to complete this PR and then follow-up with the autoselect update in a separate PR. Is this correct?

Perhaps. I'll talk to Dave about it. This is blocked by a stable release of the php-crc32 library. Looks like that is getting close.

Reply:
> @frankyn I'm sorry, I misread the benchmark script before. The third column is the number of times hash_update() is called in the test.

Thanks that source from php-crc32 helped me understand that better. So it's used to determine how many hash_update() operations occur or were able to occur given the time bounds. 

It's a little confusing, so I'm going to ignore that and focus on the Mbps instead. 

> Perhaps. I'll talk to Dave about it. This is blocked by a stable release of the php-crc32 library. Looks like that is getting close.

sgtm, thank you!

Reply:
> It's a little confusing, so I'm going to ignore that and focus on the Mbps instead.

Agreed lol!